extensions/amp-form/0.1/amp-form.js

```diff
@@ -652,6 +696,33 @@ export class AmpForm {
     return this.waitOnPromisesOrTimeout_(varSubPromises, 100);
   }
 
+  /**
+   * Call getValue() on Async Input elements, and
+   * Create hidden inputs containing their returned values
+   * @param {!Element} asyncInput
+   * @return {!Promise}
+   * @private
+   */
+  getValueForAsyncInput_(asyncInput) {
+    return asyncInput.getImpl()
+        .then(implementation => implementation.getValue())
```

Comment:
Good point @zhouyx it could make sense to add a timeout for a better submit error UX. The async variable substitution has a timeout of `100`ms, so a similar wait for the async inputs would make sense.

Reply:
Just to clarify, are we suggesting using `this.waitOnPromisesOrTimeout_` ? And the timeout is just for if the request fails silently and doesn't resolve/reject ?

What would be a good timeout value? I noticed recaptcha takes a few seconds to return a token. And any other future async input elements (google pay), I am sure could take even longer. I think 100 would be too short? But I do agree some sort of timeout would be nice.

Reply:
I believe so, @zhouyx ? 

A good value depends on how long the user should expect the response to take after clicking submit. Does the recaptcha request start at page load? If so, even if the recaptcha api response takes some time, the user probably won't submit the form until a number of seconds after the page loads. Just some sane upper-bound value that won't abandon ship too often works. 1s after submit is started? 3s? 5s? 10s?

Reply:
@cvializ recaptcha create the iframe and install service at `layoutCallback`. However the token is generating real time when user click summit, and that could take longer than 100ms. There's the external js and network involved. @torch2424 has better idea on the typical latency. 

I think @torch2424 made a good point, that different services have different typical latency. Having the same timeout value doesn't make much sense for other services we introduce later on. Instead I believe we should require every service define their own timeout value with `getValue()`. Then we can also introduce a pretty long timeout (10s?) here for safety.

Reply:
That makes sense. I agree, a timeout specific to each AsyncInput and a long safety timeout sound like good ideas.

Reply:
Awesome. I agree, lets do an upper bound of 10 seconds, and encourage other Async inputs to make a shorter timeout if appropriate? @zhouyx @cvializ 

Personally, I'd want to just use the upper bound of 10s for recaptcha. Simply because it could take even longer on 2g and NBU users and things. Though, usually in my manual testing, it takes about 1-2 seconds.